### PR TITLE
Fix #[derive] for empty structs with braces

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes https://github.com/rust-lang/rust/issues/29548

### CODE REVIEW

File: src/libsyntax_ext/deriving/debug.rs [hunk 1]
```diff
@@ -76,11 +76,16 @@ fn show_substructure(cx: &mut ExtCtxt, span: Span,
     let builder_expr = cx.expr_ident(span, builder.clone());
 
     let fmt = substr.nonself_args[0].clone();
+    let is_struct = match *substr.fields {
+        Struct(vdata, _) => vdata,
+        EnumMatching(_, v, _) => &v.node.data,
+        _ => unreachable!()
+    }.is_struct();
```
Comment (alexcrichton) [inline src/libsyntax_ext/deriving/debug.rs]: Could this be extracted at the same time that `ident` is extracted above?

**Comment by alexcrichton** [comment]
Looks good to me, thanks @petrochenkov! Just one small nit but otherwise r=me

**Comment by petrochenkov** [comment]
Updated.

**Comment by durka** [comment]
Should we back out this complexity now that `braced_empty_structs` is stable?

**Comment by petrochenkov** [comment]
Yeah, I think `derive` can be cleaned up a bit now. Tuple structs/variants still need to generate `Pat::TupleStruct`/`ExprCall`, but everything else can generate braced forms.

**Comment by Manishearth** [comment]
Please [breaking-batch] future changes to the derive machinery.

**Comment by durka** [comment]
@Manishearth what kind of changes? Changes to the public API (?) of libsyntax_ext? Or any changes at all to the expansion of `#[derive(...)]`?

**Comment by Manishearth** [comment]
The API. We're  trying to avoid libsyntax breakages that go on and break large swaths of the ecosystem. The derive machinery and the AST are both things  that do this.

**Pull request merged**